File: openmole/core/org.openmole.core.implementation/src/main/java/org/openmole/core/implementation/task/GenericTask.java
/*
 *
 *  Copyright (c) 2008, Cemagref
 *
 *  This program is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU General Public License as
 *  published by the Free Software Foundation; either version 3 of
 *  the License, or (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public
 *  License along with this program; if not, write to the Free
 *  Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston,
 *  MA  02110-1301  USA
 */
package org.openmole.core.implementation.task;

import java.io.File;
import java.util.Collection;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.TreeMap;
import java.util.TreeSet;
import java.util.logging.Level;
import java.util.logging.Logger;

import org.openmole.commons.exception.InternalProcessingError;
import org.openmole.commons.exception.UserBadDataError;
import org.openmole.commons.aspect.caching.ChangeState;
import org.openmole.core.model.data.IDataSet;
import org.openmole.core.model.execution.IProgress;
import org.openmole.core.model.resource.IResource;
import org.openmole.core.model.job.IContext;
import org.openmole.core.model.task.IGenericTask;
import org.openmole.core.model.task.annotations.Resource;
import org.openmole.core.model.task.annotations.Input;
import org.openmole.core.model.task.annotations.Output;
import org.openmole.commons.aspect.caching.SoftCachable;

import org.openmole.core.implementation.data.Data;
import org.openmole.core.implementation.data.DataSet;
import org.openmole.core.implementation.data.Parameter;
import org.openmole.core.model.data.DataModeMask;
import org.openmole.core.model.data.IVariable;
import org.openmole.core.model.data.IData;
import org.openmole.core.model.data.IParameter;
import org.openmole.core.model.data.IPrototype;

import static org.openmole.core.model.data.DataModeMask.*;
import static org.openmole.core.implementation.tools.MarkedFieldFinder.*;

public abstract class GenericTask implements IGenericTask {

    @Output
    final static public IData<Collection> Timestamps = new Data<Collection>("Timestamps", Collection.class, SYSTEM);

    @Output
    final static public IData<Throwable> Exception = new Data<Throwable>("Exception", Throwable.class, OPTIONAL, SYSTEM);

    final private String name;

    final private Map<String, IData<?>> input = new TreeMap<String, IData<?>>();;
    final private Map<String, IData<?>> output = new TreeMap<String, IData<?>>();;

    final private Set<IResource> resources = new HashSet<IResource>();

    final private Parameters parameters = new Parameters();

    public GenericTask(String name) {
        this.name = name;
    }

    /**
     * Verifies that variable specified as input are well presents in the context
     * @param context
     * @throws org.openmole.core.UserBadDataError
     * @throws org.openmole.core.InternalProcessingError
     */
    protected void verifyInput(IContext context) throws UserBadDataError, InternalProcessingError {
        for (IData<?> d : getInput()) {
            if (!d.getMode().isOptional()) {
                IPrototype<?> p = d.getPrototype();

                IVariable<?> v = context.getVariable(p.getName());
                if (v == null) {
                    throw new UserBadDataError(null, "Input data named \"" + p.getName() + "\" of type \"" + p.getType().getName() + "\" required by the task \"" + getName() + "\" has not been found");
                } else if (!p.isAssignableFrom(v.getPrototype())) {
                    throw new UserBadDataError(null, "Input data named \"" + p.getName() + "\" required by the task \"" + getName() + "\" has the wrong type: \"" + v.getPrototype().getType().getName() + "\" instead of \"" + p.getType().getName() + "\" required");
                }
            }
        }
    }

    protected void filterOutput(IContext context) throws UserBadDataError, InternalProcessingError {
        List<IVariable<?>> vars = new LinkedList<IVariable<?>>();
        for (IData<?> d : getOutput()) {
            IPrototype<?> p = d.getPrototype();
            IVariable<?> var = context.getLocalVariable(p);
            if (var != null) {
                if (p.getType().isAssignableFrom(var.getValue().getClass())) {
                    vars.add(var);
                } else {
                    Logger.getLogger(GenericTask.class.getName()).log(Level.WARNING, "Variable " + p.getName() + " of type " + p.getType().getName() + " has been found but type doesn't match : " + var.getPrototype().getType().getName() + " in task " + getName() + ".");
                }
            } else {
                if (!d.getMode().isOptional()) {
                    Logger.getLogger(GenericTask.class.getName()).log(Level.WARNING, "Variable " + p.getName() + " of type " + p.getType().getName() + " not found in output of task " + getName() + ".");
                }
            }
        }

        context.clean();
        context.setVariables(vars);
    }

    private void init(IContext context) throws UserBadDataError, InternalProcessingError {
        for (IParameter parameter : getParameters()) {
            if (parameter.getOverride() || !context.containsVariableWithName(parameter.getVariable().getPrototype())) {
                context.putVariable(parameter.getVariable().getPrototype(), parameter.getVariable().getValue());
            }
        }

        verifyInput(context);
    }

    /**
     * The main operation of the processor.
     * @param context
     * @param progress
     * @throws org.openmole.core.UserBadDataError
     * @throws org.openmole.core.InternalProcessingError
     * @throws InterruptedException
     */
    protected abstract void process(IContext context, IProgress progress) throws UserBadDataError, InternalProcessingError, InterruptedException;

    /* (non-Javadoc)
     * @see org.openmole.core.processors.ITask#run(org.openmole.core.processors.ApplicativeContext)
     */
    @Override
    public void perform(IContext context, IProgress progress) throws InternalProcessingError, UserBadDataError, InterruptedException {
        try {
            deploy();
            init(context);
            process(context, progress);
            end(context);
        } catch (InternalProcessingError e) {
            throw new InternalProcessingError(e, "Error in task " + getName());
        } catch (UserBadDataError e) {
            throw new UserBadDataError(e, "Error in task " + getName());
        }
    }

    private void end(IContext context) throws UserBadDataError, InternalProcessingError {
        filterOutput(context);
    }

    /* (non-Javadoc)
     * @see org.openmole.core.processors.ITask#addOutPrototype(org.openmole.core.data.structure.Prototype)
     */
    @Override
    public void addOutput(IPrototype prototype) {
        addOutput(new Data(prototype));
    }

    @Override
    public void addOutput(IPrototype prototype, DataModeMask... masks) {
        addOutput(new Data(prototype, masks));
    }

    @ChangeState
    @Override
    public synchronized void addOutput(IData data) {
        output.put(data.getPrototype().getName(), data);
    }

    @ChangeState
    @Override
    public synchronized void addResource(IResource resource) {
        resources.add(resource);
    }

    /* (non-Javadoc)
     * @see org.openmole.core.processors.ITask#addInPrototype(org.openmole.core.data.structure.Prototype)
     */
    @Override
    public void addInput(IPrototype prototype) {
        addInput(new Data(prototype));
    }

    @Override
    public void addInput(IPrototype prototype, DataModeMask... masks) {
        addInput(new Data(prototype, masks));
    }

    @ChangeState
    @Override
    public synchronized void addInput(IData data) {
        input.put(data.getPrototype().getName(),data);
    }

    @Override
    public void addInput(IDataSet dataSet) {
        for(IData data: dataSet) {
            addInput(data);
        }
    }

    @Override
    public void addOutput(IDataSet dataSet) {
        for(IData data: dataSet) {
            addOutput(data);
        }
    }

    @Override
    public synchronized boolean containsInput(String name) {
        if(input == null) return false;
        return input.containsKey(name);
    }

    @Override
    public boolean containsInput(IPrototype prototype) {
        return containsInput(prototype.getName());
    }

    @Override
    public synchronized  boolean containsOutput(String name) {
        if(output == null) return false;
        return output.containsKey(name);
    }

    @Override
    public boolean containsOutput(IPrototype prototype) {
        return output.containsKey(prototype.getName());
    }

    /* (non-Javadoc)
     * @see org.openmole.core.processors.ITask#getName()
     */
    @Override
    public String getName() {
        return name;
    }

    /* (non-Javadoc)
     * @see org.openmole.core.processors.ITask#getInput()
     */
    @SoftCachable
    @Override
    public IDataSet getInput() throws InternalProcessingError, UserBadDataError {
        List<IData<?>> tmpInputCache;
        tmpInputCache = new LinkedList<IData<?>>();
        if (input != null) {
            tmpInputCache.addAll(input.values());
        }

        addAllMarkedFields(this, Input.class, tmpInputCache);
        //verifyNotDuplicate(tmpInputCache, "input");

        return new DataSet(tmpInputCache);
    }

    /* (non-Javadoc)
     * @see org.openmole.core.processors.ITask#getOutput()
     */
    @SoftCachable
    @Override
    public IDataSet getOutput() throws InternalProcessingError, UserBadDataError {
        List<IData<?>> tmpOutputCache;
        tmpOutputCache = new LinkedList<IData<?>>();
        if (output != null) {
            tmpOutputCache.addAll(output.values());
        }
        addAllMarkedFields(this, Output.class, tmpOutputCache);

        return new DataSet(tmpOutputCache);
    }


    public Iterable<IData<?>> getUserInput() {
        return input.values();
    }

    public Iterable<IData<?>> getUserOutput() {
        return output.values();
    }

    @SoftCachable
    public Iterable<IResource> getResources() throws InternalProcessingError, UserBadDataError {
        Collection<IResource> resourcesCache = new HashSet<IResource>();
        if (resources != null) {
            resourcesCache.addAll(resources);
        }
        
        addAllMarkedFields(this, Resource.class, resourcesCache);
        
        return resourcesCache;
    }


    @Override
    public void deploy() throws InternalProcessingError, UserBadDataError {
        for (IResource resource : getResources()) {
            resource.deploy();
        }
    }
    
    @ChangeState
    @Override
    public void addParameter(IParameter<?> parameter) {
        parameters.put(parameter);
    }

    @Override
    public <T> void addParameter(IPrototype<? super T> prototype, T value) {
        addParameter(new Parameter<T>(prototype, value));
    }

    @Override
    public <T> void addParameter(IPrototype<? super T> prototype, T value, boolean override) {
        addParameter(new Parameter<T>(prototype, value, override));
    }

    @Override
    public Parameters getParameters() {
        return parameters;
    }


    @Override
    public String toString() {
        return getName();
    }


}

File: openmole/core/org.openmole.core.implementation/src/main/scala/org/openmole/core/implementation/task/GenericTask.scala
/*
 * Copyright (C) 2010 reuillon
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.openmole.core.implementation.task

import org.openmole.misc.exception.InternalProcessingError
import org.openmole.misc.exception.UserBadDataError
import org.openmole.core.implementation.data.Data
import org.openmole.core.implementation.data.DataSet
import org.openmole.core.implementation.data.Parameter
import org.openmole.core.model.data.DataModeMask._
import org.openmole.core.model.data.{IContext,IData,IParameter,IVariable,IPrototype,DataModeMask, IDataSet}
import org.openmole.core.model.job.ITimeStamp
import org.openmole.core.model.task.{IGenericTask,IResource}
import org.openmole.core.model.execution.IProgress
import scala.collection.immutable.TreeMap
import scala.collection.mutable.ArrayBuffer
import scala.collection.mutable.HashSet
import scala.collection.mutable.ListBuffer

object GenericTask {
  val Timestamps = new Data[ArrayBuffer[ITimeStamp]]("Timestamps#", classOf[ArrayBuffer[ITimeStamp]], Array(SYSTEM))
  val Exception = new Data[Throwable]("Exception#", classOf[Throwable], Array(OPTIONAL, SYSTEM))
}


abstract class GenericTask(val name: String) extends IGenericTask {

  private var _inputs = new TreeMap[String, IData[_]]
  private var _outputs = new TreeMap[String, IData[_]]
  private var _resources = new HashSet[IResource]
  private var _parameters = new TreeMap[String, IParameter[_]]
  
  addOutput(GenericTask.Timestamps)
  addOutput(GenericTask.Exception)
  
  protected def verifyInput(context: IContext) = {
    for (d <- inputs) {
     if (!d.mode.isOptional) {
        val p = d.prototype
        context.variable(p.name) match {
          case None => throw new UserBadDataError("Input data named \"" + p.name + "\" of type \"" + p.`type`.toString + "\" required by the task \"" + name + "\" has not been found");
          case Some(v) => if (!p.isAssignableFrom(v.prototype)) throw new UserBadDataError("Input data named \"" + p.name + "\" required by the task \"" + name + "\" has the wrong type: \"" + v.prototype.`type`.toString + "\" instead of \"" + p.`type`.toString + "\" required")
        }
      }
    }
  }

  protected def filterOutput(context: IContext): Unit = {
    val vars = new ListBuffer[IVariable[_]]
   
    for (d <- outputs) {
      val p = d.prototype
      context.variable(p) match {
        case None => 
          if (!d.mode.isOptional) throw new UserBadDataError("Variable " + p.name + " of type " + p.`type`.toString +" in not optional and has not found in output of task" + name +".")
        case Some(v) =>
          if (p.accepts(v.value)) vars += v
          else throw new UserBadDataError("Output value of variable " + p.name + " (prototype: "+ v.prototype.`type`.toString +") is instance of class '" + v.value.asInstanceOf[AnyRef].getClass + "' and doesn't match the expected class '" + p.`type`.toString + "' in task" + name + ".")
      }
    }

    context.clean
    context ++= vars
  }

  private def init(context: IContext) = {
    for (parameter <- parameters) {
      if (parameter.`override` || !context.containsVariableWithName(parameter.variable.prototype)) {
        context += parameter.variable
      }
    }

    verifyInput(context)
  }

  /**
   * The main operation of the processor.
   * @param context
   * @param progress
   */
  @throws(classOf[Throwable])
  protected def process(context: IContext, progress: IProgress): Unit
   

  /* (non-Javadoc)
   * @see org.openmole.core.processors.ITask#run(org.openmole.core.processors.ApplicativeContext)
   */
  override def perform(context: IContext, progress: IProgress) = {
    try {
      deploy
      init(context)
      process(context, progress)
      end(context)
    } catch {
      case e => throw new InternalProcessingError(e, "Error in task " + name)
    }
  }

  private def end(context: IContext) = filterOutput(context)

  override def addOutput(prototype: IPrototype[_]): this.type = addOutput(new Data(prototype))

  override def addOutput(prototype: IPrototype[_], masks: Array[DataModeMask]): this.type = addOutput(new Data(prototype, masks)); this
 
  override def addOutput(data: IData[_]): this.type =  {
    _outputs += ((data.prototype.name, data))
    this
  }
  
  override def addResource(resource: IResource): this.type = {
    _resources += resource
    this
  }
  
  override def addInput(prototype: IPrototype[_]): this.type = addInput(new Data(prototype))

  override def addInput(prototype: IPrototype[_], masks: Array[DataModeMask]): this.type = addInput(new Data(prototype, masks))

  override def addInput(data: IData[_]): this.type = {
    _inputs += ((data.prototype.name,data))
    this
  }

  override def addInput(dataSet: IDataSet): this.type = {
    for(data <- dataSet) addInput(data)     
    this
  }

  override def addOutput(dataSet: IDataSet): this.type = {
    for(data <- dataSet) addOutput(data)
    this
  }

  override def containsInput(name: String): Boolean =  _inputs.contains(name)

  override def containsInput(prototype: IPrototype[_]): Boolean = containsInput(prototype.name)

  override def containsOutput(name: String): Boolean =  _outputs.contains(name)

  override def containsOutput(prototype: IPrototype[_]): Boolean = _outputs.contains(prototype.name)

  override def inputs: IDataSet = new DataSet(_inputs)
    
  override def outputs: IDataSet = new DataSet(_outputs)
 
  @transient override lazy val userInputs: IDataSet =  new DataSet(inputs.filter(!_.mode.isSystem))
   
  @transient override lazy val userOutputs: IDataSet = new DataSet(outputs.filter(!_.mode.isSystem))

  def resources: Iterable[IResource] = _resources
 
  override def deploy = for (resource <- resources) resource.deploy   
    
  override def addParameter(parameter: IParameter[_]): this.type = {
    _parameters += ((parameter.variable.prototype.name,parameter))
    this
  }
    
  override def addParameter[T](prototype: IPrototype[T] , value: T): this.type = addParameter(new Parameter[T](prototype, value))
    
  override def addParameter[T](prototype: IPrototype[T], value: T, `override`: Boolean): this.type = addParameter(new Parameter[T](prototype, value, `override`))
    
  override def parameters: Iterable[IParameter[_]]= _parameters.values
   
  override def toString: String = name       
    
}
